code/workhorse/OOPGP.R

```diff
@@ -0,0 +1,118 @@
+# Define a class named Simulator.
+# The main method used in main.R is the simulate method. Other methods of this class and modules in other files are helper functions 
+# for the simulate method.
+
+
+library(mlegp)
+Simulator <- function(file_path_,setting_){
+  setting = setting_
+  input_path = file_path_
+  dists = NA 
+  gene_id = 15 
+  res_path = NA
+  res_name = NA
+  res_path = NA
+  data_all = NA
+  train = NA
+  train_backup = NA
+  pool = NA
+  pool_all  = NA
+  benchmark = NA
+  model = NA
+  errors = NA
+  structure(class="Simulator", environment())
+}
+
+
+source("prepare_data.R")
+
+train_simulator <- function(simulator) {
+    train = simulator$train[, c(1:14, simulator$gene_id)]
+    print("training data gene id")
+    print(dim(train))
+    num = ncol(train)
+    simulator$model = mlegp(train[,2:num-1], train[,num])
+    
+  }
+  
+
+source("update_train_pool.R")
+
+simulate_index <- function(simulator, metric_fun){
+   select_index = as.numeric()
+   for (i in 1:simulator$setting[["add"]]) {
+        metric = metric_fun(simulator)
+        index=sort(metric, decreasing = TRUE,index.return=TRUE)$ix
+   
+        select_index = c(select_index,index[1])
+        update_train_by_prediction(simulator, index[1])
+        train_simulator(simulator)
+   }
+   return (select_index)
+}
+
+
+source("screen_index_helper.R")
+source("max_dist.R")
+
+screen_index <- function(simulator) {
+  method = simulator$setting[["method"]]
+  if (method != "Random" && method != "Expert") {
+      metric_fun = switch(simulator$setting[["method"]],
+                          "EN" = EN_metric,
+                          "MI" = MI_metric,
+                          "COV"= COV_metric,
+                          "DIV"=diversity_metric,
+                          "Tan"=tanimoto_metric)
+
+      if (simulator$setting[["anti_batch"]]==1) {
+         return(simulate_index(simulator,metric_fun))
+      } else {
+         metric = metric_fun(simulator)
+         index=sort(metric, decreasing = TRUE,index.return=TRUE)$ix
+         return(index[1:simulator$setting[["add"]]])
+      }
+  } else {
+      index = seq(1,nrow(simulator$pool),1)
+      return(index[1:simulator$setting[["add"]]])
+  }
+}
+
+
+simulate <- function(simulator){
+  iter_num = simulator$setting[["iter_num"]]
+  simulator$errors = as.numeric()
+  prepare_data(simulator)
+  for (i in 1:iter_num) {
```

Comment:
very complex, nested for loop. Simplify this (maybe use `Extract Method` with good function names to help understand this.

Reply:
The nested for loop has two levels. One is ```for (i in 1:iter_num)``` and the other is ```for gene_id in gene_ids```. The meaning of each level is clear, and not so complicated, although nested for loop is not encouraged.

Reply:
I still think this should change.